src/Microsoft.ML.Recommender/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,496 @@
+            _host.Assert(yType.RawKind == DataKind.U4);
+            buffer.Get(out _m, out _n, out _k, out _p, out _q);
+            _host.Assert(_n == xType.Count);
+            _host.Assert(_m == yType.Count);
+            _host.Assert(_p.Length == _m * _k);
+            _host.Assert(_q.Length == _n * _k);
+
+            InputXType = xType;
+            InputYType = yType;
+        }
+
+        private MatrixFactorizationPredictor(IHostEnvironment env, ModelLoadContext ctx)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            // *** Binary format ***
+            // int: number of rows (m), the limit on y
+            // ulong: Minimum value of the y key-type
+            // int: number of columns (n), the limit on x
+            // ulong: Minimum value of the x key-type
+            // int: internal dimension of matrices (k)
+            // float[m * k]: the row dimension factor matrix P
+            // float[k * n]: the column dimension factor matrix Q
+
+            _m = ctx.Reader.ReadInt32();
+            _host.CheckDecode(_m > 0);
+            ulong mMin = ctx.Reader.ReadUInt64();
+            _host.CheckDecode((ulong)_m <= ulong.MaxValue - mMin);
+            _n = ctx.Reader.ReadInt32();
+            _host.CheckDecode(_n > 0);
+            ulong nMin = ctx.Reader.ReadUInt64();
+            _host.CheckDecode((ulong)_n <= ulong.MaxValue - nMin);
+            _k = ctx.Reader.ReadInt32();
+            _host.CheckDecode(_k > 0);
+
+            _p = Utils.ReadSingleArray(ctx.Reader, checked(_m * _k));
+            _q = Utils.ReadSingleArray(ctx.Reader, checked(_n * _k));
+
+            InputXType = new KeyType(DataKind.U4, nMin, _n);
+            InputYType = new KeyType(DataKind.U4, mMin, _m);
+        }
+
+        /// <summary>
+        /// Load model from the given context
+        /// </summary>
+        public static MatrixFactorizationPredictor Create(IHostEnvironment env, ModelLoadContext ctx)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            env.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(GetVersionInfo());
+            return new MatrixFactorizationPredictor(env, ctx);
+        }
+
+        /// <summary>
+        /// Save model to the given context
+        /// </summary>
+        public void Save(ModelSaveContext ctx)
+        {
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            // *** Binary format ***
+            // int: number of rows (m), the limit on y
+            // ulong: Minimum value of the y key-type
+            // int: number of columns (n), the limit on x
+            // ulong: Minimum value of the x key-type
+            // int: internal dimension of matrices (k)
+            // float[m * k]: the row dimension factor matrix P
+            // float[k * n]: the column dimension factor matrix Q
+
+            _host.Assert(_m > 0);
```

Comment:
This is public method in public class, I think it should contain Check instead of assert.

Reply:
Fixed.